### PR TITLE
fix(ferry_cache): fix regression in watch() methods 

### PR DESCRIPTION
**Pull request by knaeckeKami**
this fixes the regression I was talking about in #385 in https://github.com/gql-dart/ferry/pull/385#issuecomment-1229184574. It also fixes the test that was not testing the thing it was supposed to test.

@GP4cK If you have suggestions for better implementations, please share them

### CODE REVIEW

File: packages/ferry_cache/lib/src/cache.dart [hunk 1]
```diff
@@ -95,12 +95,41 @@ class Cache {
...
+            return;
+          }
+        } catch (err, st) {
+          if (!sc.isClosed) {
+            sc.addError(err, st);
+          }
+          //continuing immediately likely would result in an endless loop where
+          //we emit the same error over and over again
+          //so we wait for the next change
+          final isEmptyFuture = getChangeStream().first;
```
Comment (GP4cK) [inline packages/ferry_cache/lib/src/cache.dart:120]: Here should it be `.isEmpty` or `.first`?

**Comment by GP4cK** [comment]
I tried merging your branch into the one where I added the memory leak example and it seems to leak though :/

Here's the branch: https://github.com/GP4cK/ferry/tree/memory-leak-example-2

<details>
<summary>Screenshot</summary>

image

</details>

File: packages/ferry_cache/lib/src/cache.dart (hunk 1 above)
Comment (knaeckeKami) [reply packages/ferry_cache/lib/src/cache.dart:120]: probably .first, good call

**Comment by knaeckeKami** [comment]
Ah yeah true, of course this leaks as well. We need to cancel the subscription to the changeStream if the stream controller closes. .isEmpty will still listen to the stream. this should be fixed fairly  easily.

**Comment by knaeckeKami** [comment]
Now the code does not leak and supports changing data ids. 
But it's very ugly. Need to sleep over this to figure out how to clean this up.

**Comment by GP4cK** [comment]
> Now the code does not leak and supports changing data ids. But it's very ugly. Need to sleep over this to figure out how to clean this up.

I've added a draft PR based on your `todo`. PTAL: https://github.com/gql-dart/ferry/pull/403

**Comment by knaeckeKami** [comment]
Closing in favor of #403

**Pull request closed without merging**